UUID continuation

wontruefree opened the pull request:
I wanted to continue the work of UUID on crystal from https://github.com/crystal-lang/crystal/pull/2716

ysbaddaden commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,85 @@
+    UUID.byte_version @bytes[6]
+  end
+
+  # Sets variant to a specified `value`. Doesn't set variant (see `UUID#variant=(value : Variant)`).
```
Typo: "Sets **version**".

ysbaddaden commented on src/uuid/empty.cr:
```diff
@@ -0,0 +1,14 @@
+struct UUID
+  # Empty UUID.
+  @@empty_bytes = StaticArray(UInt8, 16).new { 0_u8 }
```
It should be a constant, along with some other namespaces:

```crystal
EMPTY = UUID.new(StaticArray(UInt8, 16).new(0_u8))
DNS   = UUID.new("6ba7b810-9dad-11d1-80b4-00c04fd430c8")
URL   = UUID.new("6ba7b811-9dad-11d1-80b4-00c04fd430c8")
OID   = UUID.new("6ba7b812-9dad-11d1-80b4-00c04fd430c8")
X500  = UUID.new("6ba7b814-9dad-11d1-80b4-00c04fd430c8")
```

I don't think we'd need the `#empty` method anymore —since UUID is a struct it will always be copied.

ysbaddaden commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,85 @@
+  def initialize(version : Version)
+    case version
+    when Version::V4
+      @bytes.to_unsafe.copy_from SecureRandom.random_bytes(16).pointer(16), 16
```
We can avoid a copy and directly fill `@bytes`, which also avoids a HEAP allocation :

```crystal
@bytes = uninitialized StaticArray(UInt8, 16)
SecureRandom.random_bytes(@bytes.to_slice)
```

Also, maybe we could per version factories, for a nicer API?
```crystal
def self.random
  bytes = uninitialized StaticArray(UInt8, 16)
  SecureRandom.random_bytes(bytes.to_slice)

  uuid = new(bytes)
  uuid.variant = Variant::RFC4122
  uuid.version = Version::V4
  uuid
end
```

ysbaddaden commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,85 @@
+
+  {% for v in %w(1 2 3 4 5) %}
+
+    # Returns `true` if UUID looks like V{{ v.id }}, `false` otherwise.
```
Maybe "If UUID is a Vx".

ysbaddaden commented on src/uuid/variant.cr:
```diff
@@ -0,0 +1,62 @@
+  end
+
+  # Returns UUID variant based on provided 8th `byte` (0-indexed).
+  def self.byte_variant(byte : UInt8)
```
We don't need the `byte_variant` class methods. They should be inlined into `#variant` and `#variant=`.

wontruefree replied:
I dont think I know what you mean here

ysbaddaden commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,85 @@
+    end
+  end
+
+  # Returns version based on provided 6th `byte` (0-indexed).
```
We don't need the `byte_version` class methods. They should be inlined in `#version` and `#version=`.

wontruefree replied:
I still don't think I get what you mean by this

ysbaddaden replied:
Simply that the `byte_version` class methods are only used in the version getter/setter and are thus useless. We don't need the class methods. Whatever they do should be moved to the version getter and setter.

Same for variant.

Sija commented on src/uuid.cr:
```diff
@@ -0,0 +1,63 @@
+    @bytes.to_unsafe
+  end
+
+  # Writes hyphenated format string to `io`.
```
```
`io` -> *io*
```

Sija commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,85 @@
+    UUID.byte_version @bytes[6]
+  end
+
+  # Sets Version to a specified `value`. Doesn't set variant (see `UUID#variant=(value : Variant)`).
```
```
Version -> `version`
```

Sija commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,85 @@
+
+  {% for v in %w(1 2 3 4 5) %}
+
+    # Returns `true` if UUID looks is a Vx, `false` otherwise.
```
```
Vx -> `V{{ v.id }}`
```

ysbaddaden replied:
yes, I didn't mean to drop the macro, it was great!

Sija commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,85 @@
+      variant == Variant::RFC4122 && version == RFC4122::Version::V{{ v.id }}
+    end
+
+    # Returns `true` if UUID looks is a Vx, raises `Error` otherwise.
```
ditto

Sija commented on src/uuid/string.cr:
```diff
@@ -0,0 +1,98 @@
+    end
+  end
+
+  #  Returns string in specified `format`.
```
```
`format` -> *format*
```

Sija commented on src/uuid/variant.cr:
```diff
@@ -0,0 +1,62 @@
+    UUID.byte_variant @bytes[8]
+  end
+
+  # Sets UUID variant to specified `value`.
```
```
`value` -> *value*
```

wontruefree commented:
@ysbaddaden I removed the `Random.random_bytes`

I introduced it because from what I understand SecureRandom blocks on system entropy and can it is based on IO so it can be slower and less performant but has a better source or entropy.
`Random` is a seed with a function is a lot faster but less secure.  I figured UUID is not used for cryptography so it probably does not require the performance penalty of `SecureRandom`.

ysbaddaden commented on src/uuid/string.cr:
```diff
@@ -0,0 +1,89 @@
+  #       allow non-digits. The problem it solves is that " 1".to_u8(16) is fine but if it appears inside hexstring
+  #       it's not correct and there should be stdlib function to support it, without a need to build this kind of
+  #       helpers.
+  def self.string_has_hex_pair_at!(value : String, i)
```
Reading the TODO comment, I believe this would achieve the expected result (raise if the pair isn't a strict hex number):
```crystal
value[index, 2].to_u8(16, whitespace: false, underscore: false, prefix: false)
```

ysbaddaden replied:
I meant the whole method could be replaced by this call :-)

ysbaddaden commented on src/uuid/string.cr:
```diff
@@ -0,0 +1,89 @@
+  # hexstring (ie. `89370a4ab66440c8add39e06f2bb6af6`) or URN (ie. `urn:uuid:3f9eaf9e-cdb0-45cc-8ecb-0e5b2bfb0c20`)
+  # format.
+  def decode(value : String)
+    case value.size
```
I suppose we could speed up the decode part, by using loops and counters instead of arrays, but this can be done later.

wontruefree replied:
Do you mean the `each` blocks?
Do those have a specific performance penalty?

mverzilli replied:
I think the penalty comes from the fact that each time you create an array you're causing an allocation in the heap. With loops and counters you can achieve the same results without allocations.

RX14 replied:
Can't this be simply fixed by using tuples? All the arrays here seem to be static.

In fact, they all seem to be regular: use `8.upto(23, step: 5)` for example.

luislavena replied:
@RX14 there is no `step` option for `#upto` method 😕

RX14 replied:
oh, it was `8.step(to: 23, by: 5) do |offset|`. Sorry about that, shouldn't code off the top of my head!

ysbaddaden commented on src/uuid.cr:
```diff
@@ -0,0 +1,81 @@
+  def initialize
+    bytes = StaticArray(UInt8, 16).new(0_u8)
+    bytes.to_unsafe.copy_from SecureRandom.random_bytes(16).pointer(16), 16
+    initialize bytes
```
We shouldn't call `initialize` itself. In Crystal we use a `new` factory instead. For example:

```crystal
def self.new
  bytes = ...
  new bytes
end
```

wontruefree replied:
this is generating compiler errors for me.
I am sturing to figure out what difference this makes.  If I understand that I might figure out what errors this is throwing.

ysbaddaden commented on src/uuid.cr:
```diff
@@ -0,0 +1,81 @@
+
+  # Generates UUID from static 16-`bytes`.
+  def initialize(@bytes : StaticArray(UInt8, 16))
+    initialize @bytes, Version::V4
```
Same here.

ysbaddaden commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,86 @@
+#  end
+
+  # Returns version based on provided 6th `byte` (0-indexed).
+  def self.byte_version(byte : UInt8)
```
The `.byte_version(byte)` is only ever called from `#version`. The `case` should be moved to `#version` instead.

ysbaddaden commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,86 @@
+
+  # Returns byte with encoded `version` for provided 6th `byte` (0-indexed) for known versions.
+  # For `Version::Unknown` `version` raises `ArgumentError`.
+  def self.byte_version(byte : UInt8, version : Version) : UInt8
```
Same here. The `.byte_version(byte, version)` method is only ever called from `#version=`. The `if` condition should be moved to `#version=` instead.

ysbaddaden commented on src/uuid/empty.cr:
```diff
@@ -0,0 +1,6 @@
+struct UUID
+  def self.empty
+    #initialize UUID::Version::V4, StaticArray(UInt8, 16).new(0_u8)
```
Left over comment.

wontruefree replied:
same as above

ysbaddaden commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,86 @@
+  end
+
+  # Generates RFC 4122 UUID `variant` with specified `version`.
+#  def initialize(@bytes : StaticArray(UInt8, 16), version = Version::V4)
```
Left over comment, or?

wontruefree replied:
this was from a "wip" commit I was switching computers

RX14 commented on src/uuid/rfc4122.cr:
```diff
@@ -0,0 +1,73 @@
+
+    # Returns `true` if UUID looks is a V{{ v.id }}, raises `Error` otherwise.
+    def v{{ v.id }}!
+      unless v{{ v.id }}?
```
I think making this an `if` and swapping the conditions would be a lot cleaner.

RX14 commented on src/uuid/string.cr:
```diff
@@ -0,0 +1,85 @@
+
+  def urn
+    String.new(45) do |buffer|
+      buffer.copy_from "urn:uuid:".to_unsafe, 9
```
This should use `String.build`, the risk of using unsafe pointer here isn't worth the tiny performance gain.

konovod commented on src/uuid.cr:
```diff
@@ -0,0 +1,102 @@
+  # Generates RFC 4122 v4 UUID.
+  def initialize
+    @bytes = uninitialized UInt8[16]
+    @bytes.to_unsafe.copy_from SecureRandom.random_bytes(16).pointer(16), 16
```
you can do 
```crystal
@bytes = uninitialized UInt8[16]
SecureRandom.random_bytes(@bytes.to_unsafe.to_slice(16))
```
to do the same thing without heap allocations.

RX14 replied:
Or just `SecureRandom.random_bytes(@bytes.to_slice)` actually.

wontruefree commented:
@RX14 @ysbaddaden 
Can you guys take a look at this.
I would love to get this merged.

RX14 commented:
This PR still uses `SecureRandom`, but it was removed in #4894. Please rebase this PR on master and use the examples in #4894 to convert your code.

wontruefree commented:
@RX14 So I totally touched that up this morning and forgot to push that up

RX14 commented on spec/std/uuid_spec.cr:
```diff
@@ -0,0 +1,75 @@
+    uuid.should eq("C3B46146-EB79-4E18-877B-4D46A10D1517")
+    uuid.should eq("urn:uuid:C3B46146-EB79-4E18-877B-4D46A10D1517")
+    uuid.should eq("urn:uuid:c3b46146-eb79-4e18-877b-4d46a10d1517")
+    (UUID.new).should_not eq("C3B46146-EB79-4E18-877B-4D46A10D1517")
```
What does this test?

ysbaddaden replied:
That `UUID#==` actually verifies something, and doesn't always return `true`?

RX14 commented on src/uuid.cr:
```diff
@@ -0,0 +1,277 @@
+  end
+
+  # Same as `UUID#decode(value : String)`, returns `self`.
+  def <<(value : String)
```
I'm very unsure about these methods, why would we not use the individual `variant` or `version` setters or `#decode` here? It seems like unneccesary and confusing sugar.

RX14 commented on src/uuid.cr:
```diff
@@ -0,0 +1,277 @@
+  end
+
+  # Returns `true` if `other` static 16 bytes represent the same UUID, `false` otherwise.
+  def ==(other : StaticArray(UInt8, 16))
```
I'm similarly unsure about having such weak comparisons with strings, slices, and staticarrays too. I think people really should convert manually to reduce magic.

RX14 commented on src/uuid.cr:
```diff
@@ -0,0 +1,277 @@
+      slice[4, 2].hexstring(buffer + 9)
+      slice[6, 2].hexstring(buffer + 14)
+      slice[8, 2].hexstring(buffer + 19)
+      slice[10, 6].hexstring(buffer + 24)
```
How about extracting this code which creates a slice of the to_s variant, then we can write
```
def to_s(io)
  buf = uninitialized UInt8[36]
  to_s_internal(buf.to_slice)
  io.write_utf8(buf.to_slice)
end
```

wontruefree replied:
I am not sure if that method code would be reused internally or externally.
What other method would use it?

RX14 replied:
I meant so you can provide an optimized `to_s(io)`, and an optimized `to_s`.

```cr
def to_s(io)
  buf = uninitialized UInt8[36]
  to_s_internal(buf)
  io.write_utf8(buf.to_slice)
end

def to_s
  String.new(36) do |buffer|
    to_s_internal(buffer)
    {36, 36}
  end
end

def to_s_internal(buffer)
  slice = to_slice
  buffer[8] = buffer[13] = buffer[18] = buffer[23] = 45_u8
  slice[0, 4].hexstring(buffer + 0)
  slice[4, 2].hexstring(buffer + 9)
  slice[6, 2].hexstring(buffer + 14)
  slice[8, 2].hexstring(buffer + 19)
  slice[10, 6].hexstring(buffer + 24)
end
```

currently this PR is broken because it overrides `to_s` and not `to_s(io)`.

RX14 commented on src/uuid.cr:
```diff
@@ -26,7 +24,7 @@ struct UUID
   def initialize
     @bytes = uninitialized UInt8[16]
-    @bytes.to_unsafe.copy_from SecureRandom.random_bytes(16).pointer(16), 16
+    @bytes.to_unsafe.copy_from Random::Secure.random_bytes(16).pointer(16), 16
```
You should use the new `Random::Secure.random_bytes(@bytes.to_slice)` instead.

RX14 commented on src/compiler/crystal/tools/init.cr:
```diff
@@ -102,14 +102,13 @@ module Crystal
+                     @author = "none",
+                     @email = "none",
+                     @github_name = "none",
+                     @silent = false)
```
These changes shouldn't be here (`master` always formats correctly). You must always format with the formatter built from master.

RX14 commented:
Also, shouldn't this PR include the change to remove `Random#uuid` and replace all existing UUID usage in the stdlib with this PR.

The commit history has gotten too long to be useful: please squash this into 2 commits, one to introduce `UUID` and it's specs, the second to remove `Random#uuid` and convert the stdlib.

Sija commented on src/llvm/lib_llvm_ext.cr:
```diff
@@ -14,11 +14,10 @@ lib LibLLVMExt
+                                                               builder : DIBuilder, scope : Metadata, name : Char*,
+                                                               linkage_name : Char*, file : Metadata, line : UInt,
+                                                               composite_type : Metadata, is_local_to_unit : Bool, is_definition : Bool,
+                                                               scope_line : UInt, flags : LLVM::DIFlags, is_optimized : Bool, func : LibLLVM::ValueRef) : Metadata
```
Please revert this, since it's unrelated change.

ysbaddaden commented on src/uuid.cr:
```diff
@@ -0,0 +1,251 @@
+  # Generates RFC 4122 v4 UUID.
+  def initialize
+    @bytes = uninitialized UInt8[16]
+    @bytes.to_unsafe.copy_from Random::Secure.random_bytes(16).pointer(16), 16
```
Let's make it simpler, safer (no pointers), and faster (avoids a HEAP allocation):

```crystal
@bytes = uninitialized UInt8[16]
Random::Secure.random_bytes(@bytes.to_slice)
```

ysbaddaden commented on src/uuid.cr:
```diff
@@ -0,0 +1,251 @@
+    raise ArgumentError.new "Invalid bytes length #{@bytes.size}, expected 16." unless slice.size == 16
+
+    @bytes = uninitialized UInt8[16]
+    @bytes.to_unsafe.copy_from(slice)
```
Maybe a safer (no pointers):
```crystal
@bytes = uninitialized UInt8[16]
slice.copy_to(@bytes.to_slice)
```

ysbaddaden commented on src/uuid.cr:
```diff
@@ -0,0 +1,251 @@
+
+  def initialize(version : Version)
+    @bytes = uninitialized UInt8[16]
+    @bytes.to_unsafe.copy_from Random::Secure.random_bytes(16).pointer(16), 16
```
Same as above: `SecureRandom.random_bytes(@bytes.to_slice)`

Sija replied:
Shouldn't `SecureRandom` be `Random::Secure`?

ysbaddaden replied:
Yes, habit, and juggling with 0.23 and master :)

ysbaddaden commented on src/uuid.cr:
```diff
@@ -0,0 +1,251 @@
+
+  def initialize(variant : Variant)
+    @bytes = uninitialized UInt8[16]
+    @bytes.to_unsafe.copy_from Random::Secure.random_bytes(16).pointer(16), 16
```
Same as above: `SecureRandom.random_bytes(@bytes.to_slice)`

Sija replied:
ditto

ysbaddaden commented on src/uuid.cr:
```diff
@@ -0,0 +1,251 @@
+      buffer.copy_from "urn:uuid:".to_unsafe, 9
+      (buffer + 9).copy_from to_s.to_unsafe, 36
+      {45, 45}
+    end
```
Against, let's avoid pointers, and avoid building an intermediary string representation:

```crystal
String.build(45) do |str|
  str << "urn:uuid:"
  to_s(str)
end
```

ysbaddaden commented on src/uuid.cr:
```diff
@@ -0,0 +1,251 @@
+
+  # Returns 16-byte slice.
+  def to_slice
+    Slice(UInt8).new to_unsafe, 16
```
Let's avoid yet another pointer:
```crystal
@bytes.to_slice
```

RX14 replied:
Please apply this suggestion.

wontruefree commented:
@ysbaddaden thank you for all your help and reviews.
I am sorry I am not use to the review tool so I think some of my comments are gone.
Is there anything else you can see to work on?

wontruefree commented:
@RX14 thank you so much for your help.
I think I fixed all your comments except the to_s one.
I am not sure I fully understand.

wontruefree commented:
I don't understand the CI error does anyone know what that error is?

straight-shoota commented:
Circle CI Mac builds are broken #5106 
You can just ignore that as long as Travis is green.

wontruefree commented:
sounds good thanks @straight-shoota

wontruefree commented:
@ysbaddaden @RX14 
Could you take a look and see if we can merge this?

RX14 commented on src/uuid.cr:
```diff
@@ -0,0 +1,220 @@
+  end
+
+  # Writes hyphenated format string to the *io*.
+  def to_s(io : IO)
```
Please move this overload next to the plain `to_s`.

wontruefree commented:
@RX14 thanks for your help I think I got the last 2 comments

luislavena commented on spec/std/uuid_spec.cr:
```diff
@@ -0,0 +1,150 @@
+
+describe "UUID" do
+  describe "default initialize" do
+  end
```
This `describe` block is empty. Was this intentional?

wontruefree replied:
Thanks for pointing this out.
I have updated it.

Sija commented on src/uuid.cr:
```diff
@@ -0,0 +1,220 @@
+
+  # Sets `version`. Doesn't set variant (see `#variant=`).
+  def version=(value : Version)
+    raise ArgumentError.new "Can't set unknown version." if value.unknown?
```
Please, remove the ending dot from exception message.

Sija commented on src/uuid.cr:
```diff
@@ -0,0 +1,220 @@
+    when 36 # Hyphenated
+      [8, 13, 18, 23].each do |offset|
+        if value[offset] != '-'
+          raise ArgumentError.new "Invalid UUID string format, expected hyphen at char #{offset}."
```
ditto

Sija commented on src/uuid.cr:
```diff
@@ -0,0 +1,220 @@
+        @bytes[i] = value[i * 2, 2].to_u8(16)
+      end
+    when 45 # URN
+      raise ArgumentError.new "Invalid URN UUID format, expected string starting with \":urn:uuid:\"." unless value.starts_with? "urn:uuid:"
```
ditto

Sija commented on src/uuid.cr:
```diff
@@ -0,0 +1,220 @@
+        @bytes[i] = value[offset, 2].to_u8(16)
+      end
+    else
+      raise ArgumentError.new "Invalid string length #{value.size} for UUID, expected 32 (hexstring), 36 (hyphenated) or 46 (urn)."
```
ditto

Sija commented on src/uuid.cr:
```diff
@@ -0,0 +1,220 @@
+    # Returns `true` if UUID looks is a V{{ v.id }}, raises `Error` otherwise.
+    def v{{ v.id }}!
+      unless v{{ v.id }}?
+        raise Error.new("Invalid UUID variant #{variant} version #{version}, expected RFC 4122 V{{ v.id }}.")
```
ditto

Sija commented on src/uuid.cr:
```diff
@@ -0,0 +1,220 @@
+  end
+
+  {% for v in %w(1 2 3 4 5) %}
+
```
Please, remove the empty space here, and on line 218.

wontruefree replied:
I think the line numbers may have changed.
What are you talking about I don't see extra spaces.

RX14 replied:
No, there are empty spaces here.

Sija commented on src/uuid.cr:
```diff
@@ -100,7 +100,7 @@ struct UUID
   end
 
-  # Sets `version`. Doesn't set variant (see `#variant=`).
+  # Sets `version`. Doesn't set variant (see `#variant=`)
```
Doc comments SHOULD have dots like every english sentence. Exception messages on the other hand, shouldn't. Please.

Sija commented on src/uuid.cr:
```diff
@@ -202,12 +201,12 @@ struct UUID
+    # Returns `true` if UUID looks is a V{{ v.id }}, raises `Error` otherwise
     def v{{ v.id }}!
       unless v{{ v.id }}?
         raise Error.new("Invalid UUID variant #{variant} version #{version}, expected RFC 4122 V{{ v.id }}.")
```
This commit should be reverted.

wontruefree commented:
@ysbaddaden can can this be merged?

Sija commented on spec/std/uuid_spec.cr:
```diff
@@ -0,0 +1,167 @@
+      subject.version.should eq UUID::Version::V4
+    end
+
+    it "initialize with variant" do
```
... with **slice and** variant

Sija commented on spec/std/uuid_spec.cr:
```diff
@@ -0,0 +1,167 @@
+      subject.version.should eq UUID::Version::V4
+    end
+
+    it "initialize with slice and version" do
```
... with **version**

Sija commented on spec/std/uuid_spec.cr:
```diff
@@ -0,0 +1,167 @@
+    expect_raises(ArgumentError) { UUID.new "2b1bfW06368947e59ac07c3ffdaf514c" }
+  end
+
+  it "fails when comparing to invalid strings" do
```
That looks odd, are you sure calling `==` should raise?

Sija commented on src/uuid.cr:
```diff
@@ -0,0 +1,218 @@
+  end
+
+  # Creates UUID from 16-bytes slice.
+  def initialize(slice : Slice(UInt8), variant = Variant::RFC4122, version = Version::V4)
```
`slice : Slice(UInt8)` -> `bytes : Bytes`

RX14 commented:
Sometimes I think we need to stop bikeshedding and just merge something that *good enough*. Now is one of those times.

wontruefree commented:
Thank you.

I would really like to do that because I am using this already.

Sija commented on src/random.cr:
```diff
@@ -373,34 +373,6 @@ module Random
     random_bytes(n).hexstring
   end
 
-  # Generates a UUID (Universally Unique Identifier).
```
Why there's no replacement of `Random#uuid`?
It's a handy method, why was it removed?

RX14 replied:
...because `UUID.new` doesn't use the random instance you're calling `#uuid` on, it always uses `Random::Secure`. And for good reason.

RX14 commented:
Just need 1 more review by a core team member to get this merged.

ysbaddaden commented:
I'm not sure I like removing `Random#uuid` without replacement. Random UUID doesn't have to be secure, or even random at all. There are valid scenarios for using a custom PRNG to build UUIDs from, just like there are valid use cases for other UUID types (md5, sha1, ...) with custom namespaces.

I don't like `UUID.new`: does it return a random UUID? A zero UUID? Something else? I like *explicit* factories, that is `UUID.random`, a `UUID.new(String)` factory, and a *single* initializer that requires 16-bytes and optional type and variant to apply to said bytes.

I'm not bikeshedding: I've been repeating this over and over since the beginning.

RX14 commented:
I actually agree. We should extract all the constructors which use random data to `UUID.random` and make them configurable for which `Random` instance to use - with the default being `Random::Secure` of course.

RX14 commented:
If you're tired of this PR I can take it over and finish it off for you.

wontruefree commented:
This looks like not a lot of work to close this out.

I don't understand the UUID having to be secure random.  I brought this up in the original PR I think.

This seems to be a performance penalty for not much gain.

```bash
secure 844.67k (  1.18µs) (± 6.76%) 449.65× slower
insecure 379.81M (  2.63ns) (±10.82%)        fastest
.......................
```

wontruefree commented:
@RX14 I would not mind some help.
It seems like this is very close to being done but I would very much like some help to close this out.

wontruefree commented:
It is kind of funny I picked up this PR from someone as well.

RX14 commented:
@wontruefree We don't want to use `Random` as a random number generator, we want to make it configurable *with `SecureRandom` as the default*.

RX14 commented on src/uuid.cr:
```diff
@@ -0,0 +1,217 @@
+    decode value
+
+    self.variant = variant
+    self.version = version
```
We should have an option *not* to overwrite the variant and version of the string uuid.

wontruefree replied:
This is interesting I think you have a good point.
What happens if someone sends in a variant and versions and also options?

I am not sure this follows the general conventions of Crystal but what about making the parameters have a nil or another flag.

something like `new(string, variant: :no_overwrite)`

RX14 replied:
No, by default the `variant` and `version` should be `nil` and then you use `self.variant = variant if variant`. But this shouldn't be a mutable struct.

We should have a helper method which takes a `StaticArray` and overwrites variant and version like I say, then use this before setting `@bytes` instead of having `variant=` and `version=`.

RX14 commented on src/uuid.cr:
```diff
@@ -0,0 +1,217 @@
+  # Generates RFC 4122 v4 UUID.
+  def self.random(variant = Variant::RFC4122, version = Version::V4)
+    new_bytes = uninitialized UInt8[16]
+    Random.new.random_bytes(new_bytes.to_slice)
```
The random generator should be configurable through a `random` argument, with the default being `Random::Secure`.

wontruefree commented:
@RX14 I have been mentioning since the beginning that this does not have to be secure random and I think @ysbaddaden also mentioned that.

I mean is there a reason to have it be secure random over random that is worth the 5X performance?

RX14 commented:
@wontruefree The UUID spec continuously reffers to "cryptographic-quality" random numbers.

Sorry for the delay updating this PR.

wontruefree commented:
Thanks I have been wondering why and I have not heard that yet.

wontruefree commented:
@RX14 should they be immutable?  What if someone wants to set the version or variant?

RX14 commented:
@wontruefree they create a new one, using `.new(uuid, version, variant)`.

wontruefree commented:
@RX14  This still does not cover the ignore feature I thought you mentioned.

RX14 commented:
What are we missing?

wontruefree commented:
@RX14  I would rather merge this and then write more descriptive issues on what should be changed.

I was referring to `We should have an option not to overwrite the variant and version of the string uuid `

wontruefree commented:
@RX14 @ysbaddaden 

I just wanted to check if this can be closed out.
Is there any other comments on this?

RX14 commented:
I'm going to make the changes I want right now to save time and push.

RX14 commented:
OK, pushed changes, would like another review from @crystal-lang/crystallers.

Sija commented on src/uuid.cr:
```diff
@@ -75,22 +127,6 @@ struct UUID
     end
   end
 
-  # Sets UUID variant to specified *value*.
```
Why leave out setters? Separate `UUID#variant/version=` methods look cleaner too.

RX14 replied:
Because it's a struct, and mutability is confusing. I don't see any common usecase for those functions. I did however forget to implement `UUID.new(old_uuid, variant: xxx, version: yyy)`.

wontruefree replied:
I disagree with this.
I think adding a constructor not only makes simple changes more confusing but also mean more objects will be created for no reason.

RX14 replied:
It's a struct, creating it has practically 0 cost, if not absolutely 0 const if the compiler notices you doing something like `uuid = UUID.new(uuid, variant: foo, version: foo)`. It doesn't make sense to make `UUID` mutable, it's a fixed bit pattern. The version and the variant are not metadata, they're an integral part of the bit pattern. Making a UUID object mutable is like making a string mutable, unpredictable and unnatural.

wontruefree replied:
I mean I would not mind just making this a class to get around the struct argument.
Also I do not get why immutability is a big deal?

Can you mutate this through accessing the bytes anyways?

RX14 replied:
Added `UUID.new(UUID, version, variant)`.

wontruefree commented:
@RX14 why did you remove `empty` and the ability to sent the version or the variant.
These features are ones I wanted in the PR and I think provide value.

Sija commented on src/uuid.cr:
```diff
@@ -0,0 +1,208 @@
+
+  # Returns `true` if `other` string represents the same UUID, `false` otherwise.
+  def ==(other : String)
+    self == UUID.new other
```
This line should include `rescue false` suffix since `UUID#initialize` might throw exception.

RX14 replied:
We should just remove this, in all honesty. It's magic and
- if we don't rescue exceptions using a `Hash(String | UUID, Foo)` or similar will raise.
- if we do, using a `Hash(String | UUID, Foo)` or similar will be a ddos because raising exceptions is expensive.

wontruefree replied:
This seems like a really useful method and a key feature in the code where this was abstracted.
Is there any reason why you want to remove this besides it is useful and small?

Sija replied:
@wontruefree Why not just compare `uuid.to_s == other`? Doesn't seem so bad for removing costly magic...

RX14 replied:
I removed `UUID`'s equality with `String`.

RX14 commented:
@wontruefree empty is still there, it was just moved.

wontruefree commented on src/uuid.cr:
```diff
@@ -0,0 +1,208 @@
+    buffer = uninitialized UInt8[36]
+    buffer[8] = buffer[13] = buffer[18] = buffer[23] = '-'.ord.to_u8
+
+    buffer_ptr = buffer.to_unsafe
```
I have gotten requests to explicitly remove this before.
I think having pointer manipulation is fine but I have a feeling if anyone is still following this PR they might have comments.

Also is there any reason for this or are you just changing this?

RX14 replied:
The `buffer` returned by `String.new(size, &block)` *is* a pointer. This is no less safe.

wontruefree commented:
@RX14 Good point on the `empty` method.  That to be honest was a carryover of when I picked up this PR from someone else.

wontruefree commented on src/uuid.cr:
```diff
@@ -157,12 +157,7 @@ struct UUID
   end
 
-  # Returns `true` if `other` string represents the same UUID, `false` otherwise.
-  def ==(other : String)
```
Why did you remove this?

RX14 replied:
I said exactly my reasoning above, it's magic, and costly, and unexpected.

Sija commented on src/uuid.cr:
```diff
@@ -54,6 +54,10 @@ struct UUID
     new(bytes, variant, version)
   end
 
+  def self.new(uuid : UUID, variant = nil, version = nil)
```
Why would you want to pass existing UUID with different _variant_ or _version_, what's the use case? Shouldn't this be a no-op?

RX14 replied:
No, this is a copy constructor. It creates a copy of another UUID but allows you to overwrite the variant or version, it replaces `variant=` and `version=`.

wontruefree replied:
I mean I have no problem having this in general.  I would not use it.
But I would not have deleted the code you did for this.

Sija replied:
If it's a copy constructor shouldn't the `variant` and `version` be taken from the given _uuid_?

wontruefree commented:
@RX14 hey man I wanted you help but I feel like you have not been following the PR or not listening to me as I have been working on this code.

You are blatantly ignoring me and deleting my code.
Can we get a core team member in here to review some of this?

RX14 commented:
I would be surprised if any core team member *didn't* want UUID to be immutable, but perhaps I moved too hastily and without consensus.

wontruefree commented:
Yeah I would like to see this in a PR personally there is just a lot of changes here.
But I would like to separate it out.

RX14 commented:
So you want me to create a PR instead of yours?

wontruefree commented:
@RX14 to be honest this is not how I am using UUID and I am not sure what the philosophy of the project is ie. immutability.  but whatever gets it merged at this point I am down with.  I feel like this PR is so long in the tooth I am willing to release it as a library instead of contributing upstream.

RX14 commented:
@wontruefree well, how are you using UUID?

You can do everything in this PR that you could before I made it immutable, it's just much less confusing at the cost of some rare actions being a bit more code.

RX14 commented:
This pr seems to be stalled. I suggest we review it and attempt to merge it as-is.

wontruefree commented:
Another option is I can just release this as a library and remove it from the std lib

RX14 commented:
I believe it belongs in the stdlib.

wontruefree commented:
Is there a path to get this merged?

RX14 commented:
@wontruefree yes, it just needs a positive review by another member of the core team. There might be some remaining objections but I don't see why it couldn't be merged in theory.

mverzilli commented:
@RX14 maybe it's a GitHub glitch, but what I'm seeing here is that both you and @ysbaddaden have requested changes.

@wontruefree I'll review this between today and tomorrow, thanks for the hard work and patience!

RX14 commented:
@mverzilli well I amended the PR with my own commits, so obviously I've reviewed it. As far as I know, @ysbaddaden's review is outdated.

mverzilli commented on src/uuid.cr:
```diff
@@ -0,0 +1,209 @@
+  end
+
+  {% for v in %w(1 2 3 4 5) %}
+    # Returns `true` if UUID looks is a V{{ v.id }}, `false` otherwise.
```
Did you mean "if UUID looks *as*"?

RX14 replied:
"If UUID has version Vx" would make more sense surely?

mverzilli commented on src/uuid.cr:
```diff
@@ -0,0 +1,209 @@
+      variant == Variant::RFC4122 && version == RFC4122::Version::V{{ v.id }}
+    end
+
+    # Returns `true` if UUID looks is a V{{ v.id }}, raises `Error` otherwise.
```
Ditto

mverzilli commented:
I think it looks good overall, only one question: 

Why are we constantly using `to_s` to compare UUIDs in specs? Shouldn't we directly compare UUID instances?

I understand the code will get a bit bulkier and I'm willing to merge it as is if people don't share my opinion here, but at least I'd like to know what other folks think about this before proceeding.

RX14 commented:
@mverzilli as long as the `to_s` specs properly test `to_s`, I don't see the problem. And as you said, this way is much neater.

mverzilli commented:
Thank you @wontruefree for the great work!

mirek commented:
Thank you for finishing this.

wontruefree commented:
@mirek you are welcome.  Thank you for getting it started.
